File: networklibrarydagger/src/main/java/com/colanton/federico/networklibrarydagger/NetworkConnection.java
package com.colanton.federico.networklibrarydagger;

import com.google.gson.GsonBuilder;
import com.google.gson.TypeAdapterFactory;

import javax.inject.Inject;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public final class NetworkConnection {

    private final Retrofit.Builder retrofitBuilder;

    @Inject
    public NetworkConnection(final Retrofit.Builder builder) {

        this.retrofitBuilder = builder;
    }

    private void addTypeAdapterFactories(final Retrofit.Builder builder, final TypeAdapterFactory... typeAdapterFactories) {

        if (typeAdapterFactories.length > 0) {

            for (TypeAdapterFactory factory : typeAdapterFactories) {

                GsonBuilder gsonBuilder = new GsonBuilder();
                gsonBuilder.registerTypeAdapterFactory(factory);

                GsonConverterFactory gsonConverterFactory = GsonConverterFactory.create(gsonBuilder.create());

                builder.addConverterFactory(gsonConverterFactory);
            }
        } else {

            builder.addConverterFactory(GsonConverterFactory.create());
        }
    }

    public <S> S initializeServiceInstance(final String baseUrl, final Class<S> serviceClass, final TypeAdapterFactory... typeAdapterFactories) {

        addTypeAdapterFactories(retrofitBuilder, typeAdapterFactories);

        retrofitBuilder.baseUrl(baseUrl);

        Retrofit retrofit = retrofitBuilder.build();

        return retrofit.create(serviceClass);
    }
}

File: networklibrary/src/main/java/com/colantoni/federico/networklibrary/NetworkConnection.kt
package com.colantoni.federico.networklibrary

import okhttp3.OkHttpClient
import retrofit2.CallAdapter
import retrofit2.Converter
import retrofit2.Retrofit

/**
 * Singleton class to be used as HTTP request manager.
 */
class NetworkConnection {
    private val retrofitBuilder = Retrofit.Builder()

    fun setServiceBaseUrl(baseUrl: String) {
        retrofitBuilder.baseUrl(baseUrl)
    }

    fun setOkHttpClient(okHttpClient: OkHttpClient) {
        retrofitBuilder.client(okHttpClient)
    }

    fun setCallAdapter(callAdapter: CallAdapter.Factory) {
        retrofitBuilder.addCallAdapterFactory(callAdapter)
    }

    fun setConverterAdapter(converterAdapter: Converter.Factory) {
        retrofitBuilder.addConverterFactory(converterAdapter)
    }

    /**
     * This method initializes an instance of the service you want to use.
     *
     * @param serviceClass The kind of service to be used.
     *
     * @return An initialized instance of the specified service class.
    </S> */
    fun <S> initServiceInstance(serviceClass: Class<S>): S {
        val retrofit = retrofitBuilder.build()
        return retrofit.create(serviceClass)
    }
}
